File: java/Taskie-retrofit-fragments/app/src/main/java/ada/osc/taskie/ui/tasks/adapter/TaskAdapter.java
package ada.osc.taskie.ui.tasks.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import ada.osc.taskie.R;
import ada.osc.taskie.model.Task;
import ada.osc.taskie.listener.TaskClickListener;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnLongClick;

public class TaskAdapter extends RecyclerView.Adapter<TaskAdapter.TaskViewHolder> {

    private List<Task> mTasks;
    private TaskClickListener mListener;

    public TaskAdapter(TaskClickListener listener) {
        mListener = listener;
        mTasks = new ArrayList<>();
    }

    public void updateTasks(List<Task> tasks) {
        mTasks.clear();
        mTasks.addAll(tasks);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public TaskViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_task, parent, false);
        return new TaskViewHolder(itemView, mListener);
    }

    @Override
    public void onBindViewHolder(@NonNull TaskViewHolder holder, int position) {

        Task current = mTasks.get(position);
        holder.mTitle.setText(current.getTitle());
        holder.mDescription.setText(current.getDescription());

        int color = R.color.taskPriority_Unknown;
//        switch (current.getPriority()) {
//            case LOW:
//                color = R.color.taskpriority_low;
//                break;
//            case MEDIUM:
//                color = R.color.taskpriority_medium;
//                break;
//            case HIGH:
//                color = R.color.taskpriority_high;
//                break;
//        }

        holder.mPriority.setImageResource(color);
    }

    @Override
    public int getItemCount() {
        return mTasks.size();
    }

    public void removeTask(String taskId) {
        int index = -1;

        for (Task task : mTasks) {
            if (task.getmId().equals(taskId)) {
                index = mTasks.indexOf(task);
            }
        }

        if (index != -1) {
            mTasks.remove(index);
            notifyItemRemoved(index);
        }
    }

    class TaskViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.textview_task_title)
        TextView mTitle;
        @BindView(R.id.textview_task_description)
        TextView mDescription;
        @BindView(R.id.imageview_task_priority)
        ImageView mPriority;

        public TaskViewHolder(View itemView, TaskClickListener listener) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @OnClick
        public void onTaskClick() {
            mListener.onClick(mTasks.get(getAdapterPosition()));
        }

        @OnLongClick
        public boolean onTaskLongClick() {
            mListener.onLongClick(mTasks.get(getAdapterPosition()));
            return true;
        }
    }
}

File: kotlin/Taskie/app/src/main/java/ada/osc/taskie/view/TaskAdapter.kt
package ada.osc.taskie.view

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import ada.osc.taskie.R
import ada.osc.taskie.model.Task
import ada.osc.taskie.model.TaskPriority
import kotlinx.android.synthetic.main.item_task.view.*

internal class TaskAdapter(private val taskClickListener: TaskClickListener) : RecyclerView.Adapter<TaskAdapter.TaskViewHolder>() {

    private val tasks = mutableListOf<Task>()

    fun updateTasks(tasks: List<Task>) {
        this.tasks.clear()
        this.tasks.addAll(tasks)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TaskViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.item_task, parent, false)
        return TaskViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: TaskViewHolder, position: Int) {
        holder.displayData(tasks[position], taskClickListener)
    }

    override fun getItemCount(): Int {
        return tasks.size
    }

    internal inner class TaskViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun displayData(task: Task, listener: TaskClickListener) = with(itemView) {
            taskTitle.text = task.title
            taskDescription.text = task.description

            taskPriority.setImageResource(when (task.priority) {
                TaskPriority.LOW -> R.color.taskpriority_low
                TaskPriority.MEDIUM -> R.color.taskpriority_medium
                TaskPriority.HIGH -> R.color.taskpriority_high
            })

            setOnClickListener { listener.onClick(task) }
            setOnLongClickListener{
                listener.onLongClick(task)
                true
            }
        }
    }
}
